File: backend/app/models/alert.py
"""알림 시스템 Pydantic 모델."""

from datetime import datetime

from pydantic import BaseModel, field_validator


class PriceAlertCreateRequest(BaseModel):
    """가격 알림 생성 요청."""

    ticker: str
    company_name: str | None = None
    alert_type: str  # TARGET_PRICE | STOP_LOSS
    trigger_price: float
    memo: str | None = None

    @field_validator("ticker")
    @classmethod
    def normalize_ticker(cls, v: str) -> str:
        return v.strip().upper()

    @field_validator("alert_type")
    @classmethod
    def validate_alert_type(cls, v: str) -> str:
        allowed = {"TARGET_PRICE", "STOP_LOSS"}
        if v not in allowed:
            raise ValueError(f"alert_type은 {allowed} 중 하나여야 합니다.")
        return v

    @field_validator("trigger_price")
    @classmethod
    def validate_trigger_price(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("trigger_price는 0보다 커야 합니다.")
        return v


class DeleteResponse(BaseModel):
    """삭제 결과 응답."""

    success: bool = True
    message: str


class PriceAlertResponse(BaseModel):
    """DB price_alerts 테이블 1:1 대응."""

    id: str
    user_id: str
    ticker: str
    company_name: str | None = None
    alert_type: str  # TARGET_PRICE | STOP_LOSS | DAILY_CHANGE
    trigger_price: float
    current_price: float | None = None
    is_triggered: bool = False
    triggered_at: datetime | None = None
    memo: str | None = None
    created_at: datetime


class PriceAlertsListResponse(BaseModel):
    """사용자 가격 알림 목록."""

    alerts: list[PriceAlertResponse]
    total: int


class AlertCheckResult(BaseModel):
    """가격 알림 체크 결과 요약."""

    checked_count: int
    triggered_count: int
    notified_count: int
    failed_tickers: list[str]
    checked_at: datetime


class RiskAlertResult(BaseModel):
    """리스크 조건 체크 결과."""

    vix_alert: bool = False
    vix_value: float | None = None
    geopolitical_alert: bool = False
    high_urgency_count: int = 0
    currency_alert: bool = False
    usd_krw_change_pct: float | None = None
    notifications_sent: int = 0
    checked_at: datetime

File: src/lib/api/alert.ts
/**
 * Alert(가격 알림) API 클라이언트.
 */

import { apiFetch } from "./client";

// ─── 타입 ───

export interface PriceAlert {
  id: string;
  user_id: string;
  ticker: string;
  company_name: string | null;
  alert_type: string; // TARGET_PRICE | STOP_LOSS
  trigger_price: number;
  current_price: number | null;
  is_triggered: boolean;
  triggered_at: string | null;
  memo: string | null;
  created_at: string;
}

interface PriceAlertsListResponse {
  alerts: PriceAlert[];
  total: number;
}

// ─── API 함수 ───

/** 내 가격 알림 목록 */
export async function listAlerts(): Promise<PriceAlert[]> {
  const res = await apiFetch<PriceAlertsListResponse>("/alert/my");
  return res.alerts;
}

/** 가격 알림 생성 */
export async function createAlert(data: {
  ticker: string;
  company_name?: string;
  alert_type: string;
  trigger_price: number;
  memo?: string;
}): Promise<PriceAlert> {
  return apiFetch<PriceAlert>("/alert/", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/** 가격 알림 삭제 */
export async function deleteAlert(
  alertId: string,
): Promise<{ success: boolean; message: string }> {
  return apiFetch(`/alert/${alertId}`, { method: "DELETE" });
}

// ─── 관리자 전용 ───

export interface AlertCheckResult {
  checked_count: number;
  triggered_count: number;
  notified_count: number;
  failed_tickers: string[];
  checked_at: string;
}

export interface RiskAlertResult {
  vix_alert: boolean;
  vix_value: number | null;
  geopolitical_alert: boolean;
  high_urgency_count: number;
  currency_alert: boolean;
  usd_krw_change_pct: number | null;
  notifications_sent: number;
  checked_at: string;
}

/** [관리자] 가격 알림 수동 확인 */
export async function checkPriceAlerts(): Promise<AlertCheckResult> {
  return apiFetch<AlertCheckResult>("/alert/check", { method: "POST" });
}

/** [관리자] 리스크 알림 수동 확인 (VIX·환율·지정학) */
export async function checkRiskAlerts(): Promise<RiskAlertResult> {
  return apiFetch<RiskAlertResult>("/alert/risk-check", { method: "POST" });
}
